Add spot instance support in AWSMachinePool

Ankitasw opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This PR adds support for spot instance in AWSMachinePool. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes Part of #3522 

**Checklist**:

- [x] squashed commits
- [x] includes documentation
- [x] adds unit tests
- [x] adds or updates e2e tests

Ankitasw commented:
/test pull-cluster-api-provider-aws-e2e

Ankitasw commented:
This PR is ready for review.

sedefsavas commented:
Thanks for the changes.

Double checking: have you confirmed that CAPI MachinePool test validates both MachinePools created? Plus, does it actually use spot instances as expected (could be seen from AWS console)?

sedefsavas commented on docs/triage-party/triage-party-deployment.go:
```diff
@@ -18,16 +18,14 @@ package main
 import (
 	"fmt"
-
+	
```
This file should not change. 
Also [docs/proposal/20200506-single-controller-multitenancy-flow.svg](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3563/files#diff-189069f4133909854878c049610da1cbc5dd62c8f94b830887e200b86aee8d74) changes above.

Ankitasw replied:
I think some make target messed it up, I ll revert it.

sedefsavas commented on pkg/cloud/services/ec2/launchtemplate.go:
```diff
@@ -522,3 +524,33 @@ func (s *Service) getFilteredSecurityGroupID(securityGroup infrav1.AWSResourceRe
+	// - If the spot request cannot immediately be fulfilled, it will not be created
+	// This behaviour should satisfy the 1:1 mapping of Machines to Instances as
+	// assumed by the Cluster API.
+	spotOptions.SetInstanceInterruptionBehavior(ec2.InstanceInterruptionBehaviorTerminate)
+	spotOptions.SetSpotInstanceType(ec2.SpotInstanceTypeOneTime)
```
Persistent option is not applicable for ec2 auto scaling, you can see the message from AWS console when you manually try to set the spot instance type, there is only one option.
`Persistent: not applicable for EC2 auto scaling`
`Specify a persistent request so that interrupted Spot Instances are requested again. If you do not specify a request type, EC2 makes a one-time request by default. Persistent requests are only supported when interruption behavior is set to either hibernate or stop.`

So we don't need to set spot instance as default is one-time.

sedefsavas replied:
Interruption behavior, again in console, it won't allow to set. The message there:
`For one-time requests, only terminate is valid. `

Default for SetInstanceInterruptionBehavior is terminate so no need to set here.

sedefsavas replied:
Comments need to cover these.
I don't think 1:1 matching is related to these settings, for each instance there will be a new Machine object when MachinePool Machine proposal is implemented.

sedefsavas replied:
In summary, not all the options are valid for ec2 auto scaling.

Ankitasw replied:
Thanks @sedefsavas , will take care of this.

Ankitasw commented:
> Double checking: have you confirmed that CAPI MachinePool test validates both MachinePools created?

Good point @sedefsavas , I checked earlier in logs only if machine pool is created, but looks like its not creating 2nd machine pool. I will debug locally what's going wrong.

Ankitasw commented:
> does it actually use spot instances as expected (could be seen from AWS console)?

@sedefsavas Yes it does. 
Also, I have addressed all the comments in recent commit. PTAL

Ankitasw commented:
/test pull-cluster-api-provider-aws-e2e

Skarlso commented on docs/book/src/topics/spot-instances.md:
````diff
@@ -49,4 +49,22 @@ spec:
+       maxPrice: ""
+```
+
+> **IMPORTANT WARNING**: The experimental feature `AWSMachinePool` now supports using spot instances, but graceful shutdown of the machines in `AWSMachinePool` is not supported and has to be handled externally by users.
````
```suggestion
> **IMPORTANT WARNING**: The experimental feature `AWSMachinePool` supports using spot instances, but a graceful shutdown of the machines in `AWSMachinePool` is not supported and has to be handled externally by users.
```

The `now` is not really needed. It makes sense when you write it with the context that it has just been enabled, but when reading it first on the website, it adds a time scale without a frame of reference, if that makes sense? :D

Ankitasw replied:
Yes that makes sense 😄

Skarlso commented on exp/api/v1beta1/types.go:
```diff
@@ -113,6 +113,9 @@ type AWSLaunchTemplate struct {
 	// +optional
 	AdditionalSecurityGroups []infrav1.AWSResourceReference `json:"additionalSecurityGroups,omitempty"`
+
+	// SpotMarketOptions option for configuring AWSMachinePool instances to be run using AWS Spot instances.
```
```suggestion
	// SpotMarketOptions options for configuring AWSMachinePool instances to be run using AWS Spot instances.
```

Skarlso commented on pkg/cloud/services/ec2/launchtemplate.go:
```diff
@@ -522,3 +524,28 @@ func (s *Service) getFilteredSecurityGroupID(securityGroup infrav1.AWSResourceRe
+	// Persistent requests are only supported when interruption behavior is set to either hibernate or stop.
+	// By default, interruption behavior is `terminate`, so no need to explicitly specify that here as well.
+
+	maxPrice := spotMarketOptions.MaxPrice
+	if maxPrice != nil && *maxPrice != "" {
+		spotOptions.SetMaxPrice(*maxPrice)
+	}
```
```suggestion
	if maxPrice := aws.StringValue(spotMarketOptions.MaxPrice); maxPrice != "" {
		spotOptions.SetMaxPrice(maxPrice)
	}
```

Since you aren't using maxPrice we could use StringVaule like this. WDYT? :)

Ankitasw replied:
Sounds good to me

Skarlso commented on pkg/cloud/services/ec2/launchtemplate.go:
```diff
@@ -522,3 +524,28 @@ func (s *Service) getFilteredSecurityGroupID(securityGroup infrav1.AWSResourceRe
+	}
+
+	launchTemplateInstanceMarketOptionsRequest := &ec2.LaunchTemplateInstanceMarketOptionsRequest{}
+	launchTemplateInstanceMarketOptionsRequest.SetMarketType(ec2.MarketTypeSpot)
```
I'll be honest, I don't much like the set methods. They are sneaky. :D They return the options, but they operate on a pointer value. But if you don't use the return value, there can be surprises with slices.

I'm not sure about them, they are generated so I usually just stick to applying the value as is to the struct.

This is absolutely a personal opinion. :)

Ankitasw replied:
I just followed whatever we have today for AWSMachine 😄 although I am ok to use both ways, but can you explain how SET methods breaks rules, would be great if you have some references.

Skarlso replied:
Nah, I just wanted to rant a bit. :DD It's fine, since the rest of the whole thing uses it, so you are right that we probably shouldn't mix semantics. :) Otherwise, it's just confusing to which one to use at which point. Also, I know it's convenient because it needs the actual value instead of the pointer.

It's finneeee. Ignore me. :DDD

Ankitasw replied:
That's completely alright 😄

Skarlso commented:
FWIW LGTM :))))))

sedefsavas commented:
@richardcase WDYT about backporting this to v1.5.0?

richardcase commented:
> WDYT about backporting this to v1.5.0?

I think this would be good. So:

/cherry-pick release-1.5

richardcase commented on test/e2e/data/infrastructure-aws/kustomize_sources/machine-pool/spot-instance-machine-pool.yaml:
```diff
@@ -0,0 +1,45 @@
+    instanceType: "${AWS_CONTROL_PLANE_MACHINE_TYPE}"
+    iamInstanceProfile: "nodes.cluster-api-provider-aws.sigs.k8s.io"
+    sshKeyName: "${AWS_SSH_KEY_NAME}"
+    spotMarketOptions:
```
Should this be:

```yaml
spotMarketOptions: {}
```

Ankitasw replied:
I referred the existing test where we have specified nil maxPrice

> spotMarketOptions:
>      maxPrice: ""

Ankitasw replied:
Can this be relapced with 
```yaml
spotMarketOptions: {}
```

sedefsavas replied:
I think so the check is for non-nil options.

Ankitasw replied:
@sedefsavas so shall we keep it as is?

richardcase commented on test/e2e/data/infrastructure-aws/kustomize_sources/machine-pool/spot-instance-machine-pool.yaml:
```diff
@@ -0,0 +1,45 @@
+apiVersion: cluster.x-k8s.io/v1beta1
```
I don't see this template being used in any e2e tests. Should there also be a change in `e2e_conf.yaml` and then use the flavour in a test.

Or have i missed its usage?

Ankitasw replied:
this is used in kustomization yaml itself https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/1e938c88ebe75470f11e0159fc24d8f8212da03b/test/e2e/data/infrastructure-aws/kustomize_sources/machine-pool/kustomization.yaml#L4
We have not created separate e2e test as CAPI MachinePool test expect only one flavor, so we have combined both in 1 test.

richardcase replied:
Got you, thanks @Ankitasw.

Do we want to combine the tests long term?

Ankitasw replied:
I knowingly kept these two as separate tests, as both are for different use cases.

sedefsavas replied:
> Good point @sedefsavas , I checked earlier in logs only if machine pool is created, it showed created in logs, but looks like its not creating 2nd machine pool in aws console

@Ankitasw this is resolved now? 
CAPI MP test creates both the MPs?

Ankitasw replied:
yes this is resolved.

richardcase reviewed (commented):
This looks good to me. A couple of questions around the e2e template.

Ankitasw commented:
@sedefsavas @richardcase  gentle reminder for review!!

Skarlso commented:
👋 
/lgtm

sedefsavas commented on pkg/cloud/services/ec2/launchtemplate.go:
```diff
@@ -522,3 +524,27 @@ func (s *Service) getFilteredSecurityGroupID(securityGroup infrav1.AWSResourceRe
+	spotOptions := &ec2.LaunchTemplateSpotMarketOptionsRequest{}
+
+	// Persistent option is not applicable EC2 autoscaling. If we do not specify a request type, EC2 makes a one-time request by default.
+	// Persistent requests are only supported when interruption behavior is set to either hibernate or stop.
```
Since we don't use persistent requests, IMO no need to add its interruption behaviour here.

sedefsavas reviewed (commented):
Just few comment fixes, otherwise LGTM as well.

Skarlso commented:
/lgtm with new changes. Nice spots on readability Sedef. :)

sedefsavas commented:
/lgtm
/approve

fyi @Skarlso, lgtm and other commands need to be on their own line. Comments added after those commands makes them not work.
